### PR TITLE
autoconfig WSC add M1/M2 parameters in agent and controller

### PR DESCRIPTION
**Pull request by tomereli**
## Overview
Add M1 M2 parameters set for the WSC messages sent from the controller and agent (and vice versa) as part of the AP autoconfiguration flow.
Currently all agent M1 parameters are dummy, and all the controller's M2 are either based on the agent or dummy as well.
With this, we can see the full M1 and M2(s) sent and received correctly using wireshark.

Related Task - #71 

## Changes summary
- Fix vendor extention yaml typo
- Fix a bug where the 5GHz slave is set to 2.4Ghz in dummy mode
- Port a fix to a bug in the backhaul manager which causes M1 to be sent twice by the backhaul manager slave
- Add helper functions to support setting and reading WSC attributes which are not covered by TLVF
- Set all M1 parameters in the agent autoconfig WSC message sent to the controller during autoconfig.
- Set all M2 parameters in the controller autoconfig WSC message based on the M1 parameters received from the agent

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

### CODE REVIEW

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 1]
```diff
@@ -458,6 +458,12 @@ bool main_thread::finalize_slaves_connect_state(bool fConnected,
                 continue;
             }
```
Comment (morantr) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Regarding This commit which fix double sending of backhaul connected notification.
I already had a fix for it on Intel Bitbucket. See commit 22956a4d3a6, 23 April 2019, and lock for:
"soc->backhaul_operational_on_registration".

File: common/beerocks/tlvf/AutoGenerated/include/beerocks/tlvf/beerocks_wsc.h [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ */
+
+#ifndef _TLVF_WSC_WSC_ATTRIBUTES_HELPERS_H_
+#define _TLVF_WSC_WSC_ATTRIBUTES_HELPERS_H_
+
+#include <beerocks/bcl/beerocks_utils.h>
+#include <tlvf/WSC/WSC_Attributes.h>
+
+namespace WSC {
+    enum class eBssType : uint8_t {
```
Comment (morantr) [inline common/beerocks/tlvf/AutoGenerated/include/beerocks/tlvf/beerocks_wsc.h]: Why enum class and not regular enum?

File: agent/src/beerocks/slave/son_slave_thread.cpp [hunk 3]
```diff
@@ -3678,23 +3679,22 @@ bool slave_thread::slave_fsm(bool &call_slave_select)
             break;
         }
 
-        auto apconfHeader =
-            cmdu_tx.create(0, ieee1905_1::eMessageType::AP_AUTOCONFIGURATION_WSC_MESSAGE);
-        auto tlvAp = cmdu_tx.addClass<wfa_map::tlvApRadioBasicCapabilities>();
-        if (tlvAp == nullptr) {
+        auto c  = cmdu_tx.create(0, ieee1905_1::eMessageType::AP_AUTOCONFIGURATION_WSC_MESSAGE);
```
Comment (morantr) [inline agent/src/beerocks/slave/son_slave_thread.cpp]: When if c == nullptr?
second, I don't think using 'c' as a name is good idea. It is not meaningful. 
Please consider to change it to "cmdu_header".

File: agent/src/beerocks/slave/son_slave_thread.cpp [hunk 4]
```diff
@@ -3678,23 +3679,22 @@ bool slave_thread::slave_fsm(bool &call_slave_select)
...
+        auto ap = cmdu_tx.addClass<wfa_map::tlvApRadioBasicCapabilities>();
```
Comment (morantr) [inline agent/src/beerocks/slave/son_slave_thread.cpp]: "ap" is not meaningful name as well.
Please consider to change it to "tlvApRadioBasicCapabilities".

File: agent/src/beerocks/slave/son_slave_thread.cpp [hunk 5]
```diff
@@ -3678,23 +3679,22 @@ bool slave_thread::slave_fsm(bool &call_slave_select)
...
+        if (ap == nullptr) {
             LOG(ERROR) << "Error creating TLV_AP_RADIO_BASIC_CAPABILITIES";
             return false;
         }
         std::copy_n(network_utils::mac_from_string(config.radio_identifier).oct, sizeof(sMacAddr),
```
Comment (morantr) [inline agent/src/beerocks/slave/son_slave_thread.cpp]: This line needs to be changed now to:
ap->radio_uid() = network_utils::mac_from_string(config.radio_identifier);

File: agent/src/beerocks/slave/son_slave_thread.cpp [hunk 6]
```diff
@@ -3708,53 +3708,52 @@ bool slave_thread::slave_fsm(bool &call_slave_select)
...
                 LOG(ERROR) << "add_operating_classes_info_list failed";
                 return false;
             }
         }
 
         // All attributes which are not explicitely set below are set to
         // default by the TLV factory, see WSC_Attributes.yml
-        auto tlvWscM1 = cmdu_tx.addClass<ieee1905_1::tlvWscM1>();
-        if (tlvWscM1 == nullptr) {
+        auto m1 = cmdu_tx.addClass<ieee1905_1::tlvWscM1>();
```
Comment (morantr) [inline agent/src/beerocks/slave/son_slave_thread.cpp]: why did you change the name? we usually use the name of the tlv

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 1 above)
Comment (tomereli) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: ok, I didn't know. I will take your change instead - thanks.

File: common/beerocks/tlvf/AutoGenerated/include/beerocks/tlvf/beerocks_wsc.h (hunk 2 above)
Comment (tomereli) [reply common/beerocks/tlvf/AutoGenerated/include/beerocks/tlvf/beerocks_wsc.h]: no real reason - I will change to regular enum.

File: agent/src/beerocks/slave/son_slave_thread.cpp (hunk 3 above)
Comment (tomereli) [reply agent/src/beerocks/slave/son_slave_thread.cpp]: Fixed.
Regarding using short names - generally I agree, but in this case we only use the return pointer to check if `create` succeeded. In such cases I prefer to keep the line short by using a short variable since it is not used other than error checking anyway. I encourage you to do the same, will prevent clang-format from "uglying" the code.

File: agent/src/beerocks/slave/son_slave_thread.cpp (hunk 4 above)
Comment (tomereli) [reply agent/src/beerocks/slave/son_slave_thread.cpp]: This name is way too long.. I will change to radio_basic_caps - we don't care it is a TLV, we just use the class.

File: agent/src/beerocks/slave/son_slave_thread.cpp (hunk 5 above)
Comment (tomereli) [reply agent/src/beerocks/slave/son_slave_thread.cpp]: Nice one, thanks. done.

File: common/beerocks/bwl/dummy/base_wlan_hal_dummy.cpp [hunk 7]
```diff
@@ -88,6 +88,7 @@ base_wlan_hal_dummy::base_wlan_hal_dummy(HALType type, std::string iface_name, b
     }
 
     // Initialize the FSM
+    refresh_radio_info();
```
Comment (simonbaren) [inline common/beerocks/bwl/dummy/base_wlan_hal_dummy.cpp]: Maybe you should keep the logic in other BWL types - refresh_radio_info on attach instead...

File: agent/src/beerocks/slave/son_slave_thread.cpp (hunk 6 above)
Comment (tomereli) [reply agent/src/beerocks/slave/son_slave_thread.cpp]: In this case m1 is enough - no need to mention it is a TLV.

File: common/beerocks/tlvf/AutoGenerated/include/beerocks/tlvf/beerocks_wsc.h (hunk 2 above)
Comment (simonbaren) [reply common/beerocks/tlvf/AutoGenerated/include/beerocks/tlvf/beerocks_wsc.h]: Do you plan to use this enum without `WSC::eBssType::`?
If not - keep the enum class

File: agent/src/beerocks/slave/son_slave_thread.cpp (hunk 3 above)
Comment (simonbaren) [reply agent/src/beerocks/slave/son_slave_thread.cpp]: `if (!cmdu_tx.create(0, ieee1905_1::eMessageType::AP_AUTOCONFIGURATION_WSC_MESSAGE)) {`
same length ~

File: agent/src/beerocks/slave/son_slave_thread.cpp [hunk 8]
```diff
@@ -3678,23 +3679,22 @@ bool slave_thread::slave_fsm(bool &call_slave_select)
             break;
         }
 
-        auto apconfHeader =
-            cmdu_tx.create(0, ieee1905_1::eMessageType::AP_AUTOCONFIGURATION_WSC_MESSAGE);
-        auto tlvAp = cmdu_tx.addClass<wfa_map::tlvApRadioBasicCapabilities>();
-        if (tlvAp == nullptr) {
+        auto c  = cmdu_tx.create(0, ieee1905_1::eMessageType::AP_AUTOCONFIGURATION_WSC_MESSAGE);
+        auto ap = cmdu_tx.addClass<wfa_map::tlvApRadioBasicCapabilities>();
+        if (ap == nullptr) {
```
Comment (simonbaren) [inline agent/src/beerocks/slave/son_slave_thread.cpp]: for shared_ptr it is better to check for `if (!ap) {`

File: agent/src/beerocks/slave/son_slave_thread.cpp [hunk 9]
```diff
@@ -3708,53 +3708,52 @@ bool slave_thread::slave_fsm(bool &call_slave_select)
...
+            hostap_params.iface_is_5ghz ? WSC::WSC_RF_BAND_5GHZ : WSC::WSC_RF_BAND_2GHZ;
+        WSC::set_vendor_extensions(m1->M1Frame().vendor_extensions_attr,
+                                   WSC::eBssType::FRONTHAUL_BSS);
+        WSC::set_primary_device_type(m1->M1Frame().primary_device_type_attr);
+        // TODO: M1 should also have values for:
+        // enrolee_nonce_attr -> to be added by encryption
+        // public_key_attr -> to be added by encryption
+
+        auto vs = cmdu_tx.add_vs_tlv(ieee1905_1::tlvVendorSpecific::eVendorOUI::OUI_INTEL);
+        if (vs == nullptr) {
```
Comment (simonbaren) [inline agent/src/beerocks/slave/son_slave_thread.cpp]: shared_ptr should be checked with `if (!ap) {`

File: common/beerocks/tlvf/AutoGenerated/include/beerocks/tlvf/beerocks_wsc.h (hunk 2 above)
Comment (tomereli) [reply common/beerocks/tlvf/AutoGenerated/include/beerocks/tlvf/beerocks_wsc.h]: Yes - I will change it to use uint8_t so better to use simple enum as the values are only bitmasks.

File: common/beerocks/bwl/dummy/base_wlan_hal_dummy.cpp (hunk 7 above)
Comment (tomereli) [reply common/beerocks/bwl/dummy/base_wlan_hal_dummy.cpp]: Done

File: common/beerocks/bwl/dummy/base_wlan_hal_dummy.cpp [hunk 10]
```diff
@@ -88,15 +88,18 @@ base_wlan_hal_dummy::base_wlan_hal_dummy(HALType type, std::string iface_name, b
     }
```
Comment (simonbaren) [inline common/beerocks/bwl/dummy/base_wlan_hal_dummy.cpp]: change the commit message accordingly

**Review by simonbaren — APPROVED** [review]

**Comment by tomereli** [comment]
GW-Repeater-M1-M2-operational

**Comment by tomereli** [comment]
> GW-Repeater-M1-M2-operational

What we see in this screenshot is GW+Repeater operational in dummy mode and M1 + M2 sniffed in Wireshark.
The reason we see 4 M2 messages is because we set the maximum number of BSS in the M1 message for each radio to be 4 (hardcoded).
